plugins/optimization-detective/class-od-html-tag-processor.php

```diff
@@ -160,7 +179,32 @@ final class OD_HTML_Tag_Processor {
 	 * @param string $html HTML to process.
 	 */
 	public function __construct( string $html ) {
-		$this->processor = new WP_HTML_Tag_Processor( $html );
+
+		$this->processor = new class( $html ) extends WP_HTML_Tag_Processor {
```

Comment:
This seems a bit fragile to me. Can we make it an actual named class?

Reply:
Fragile how? It's what @dmsnell suggested so I follow his lead for the use of the anonymous class.

Reply:
I've invited @dmsnell to the Performance team on GitHub for proper approval granting 😄

Reply:
I'm not sure about what kind of fragility we're concerned about here. Extending `WP_HTML_Tag_Processor` provides guarantees about the behavior of the underlying methods.

The anonymous class was more or less an example I tossed out as a convenient way to create or expose risky behaviors from the HTML API without exporting that to the outside world. It creates these methods but they shouldn't be instantiable or callable from outside, and won't show up in documentation. Also they are convenient if your styling rules forbid having more than one class in the same file.

Reply:
I'm just questioning why we use an anonymous class here. We haven't done that anywhere, even with other classes we consider private. So it's a new pattern and I'm not sure we want to go that route, as it would lead to more such cases in the future.

As @dmsnell says, the anonymous class was just "a convenient way". Changing it to a regular class will probably take just a few minutes, but follow our patterns established so far.

I just don't see any good reason for an anonymous class, other than it was very quick to throw in here and we don't have to think about naming. We can just mark it as `@access private` like we do with other classes, then it doesn't matter if someone else tries to use it - that's on them, and there's no support or guarantees for it.

Reply:
It makes sense to me to use an anonymous class here since this is the only place where this class is being used. Making it a regular class would mean yet another class in the global namespace that isn't being used anywhere except this one place.

Reply:
fun fact I learned recently: anonymous classes _aren't_ anonymous 😄

_thankfully_, they are compiled up-front, given a unique name, and behave like regular classes. if you return one from a function then all the returned values will be of the same "anonymous" class.

```php
function make() {
	return new class () {
		public function me() {
			echo "I am no sweet unique butterfly 😭";
		}
	}
}

$a = make();
$b = make();
true === ( get_class( $a ) === get_class( $b ) );
true === ( $a instanceof $b );
true === ( $b instanceof $a );
```

this is unrelated to the current discussion, but I found that interesting when I learned.

Reply:
@westonruter 
> It makes sense to me to use an anonymous class here since this is the only place where this class is being used. Making it a regular class would mean yet another class in the global namespace that isn't being used anywhere except this one place.

IMO that's not a good argument against using a regular class. For example, the `OD_HTML_Tag_Processor` class is also only instantiated and used in one place in the plugin code, so I don't see the difference. Also, as mentioned before, it's a new pattern and I'm not sure it's one we want to encourage.

Reply:
@felixarntz The other thing is that the use of `WP_HTML_Tag_Processor` should eventually be replaced with `WP_HTML_Processor` once it implements support for the remaining tags. This class will eventually support node insertion itself, so what is currently an anonymous class here should eventually be removed. IMO just because we haven't used a pattern before (here because it's a new language feature in PHP 7) isn't a good argument against using it. But if you feel strongly I'll do the work to move it to a regular class.

Reply:
I'd prefer to do it in a class. Once we no longer need it, we can deprecate it, or even remove if marked private.

Reply:
I'd prefer not 😄 

I guess you feel strongly then. I'll do it.

Reply:
OK, this is done.

Reply:
Thanks @westonruter, I think this is in line with how we've been writing classes so far. FWIW there are no anonymous classes in WP core either, and since this is a feature plugin I think it makes sense to follow core's patterns too.

Reply:
Right, but this was intended to anticipate `WP_Tag_Processor` having such node insertion functionality. By the time this is proposed for core I was expecting the anonymous class would no longer be needed. Anyway, let's move on 😄

Reply:
might be good to consider anonymous classes at times when we have something dangerous like this that we don't want to expose. it's unfortunate we have no way of hiding things otherwise.